src/ssl.c

```diff
@@ -12459,7 +12460,7 @@ WOLFSSL_SESSION* wolfSSL_get_session(WOLFSSL* ssl)
 {
     WOLFSSL_ENTER("SSL_get_session");
     if (ssl)
-        return wolfSSL_GetSession(ssl, NULL, 1);
+        return &ssl->session;
```

Comment:
This doesn't work. `wolfSSL_get_session` is called... WOLFSSL object can be free'd between resumption.

Reply:
That is correct because `wolfSSL_get_session` should return a reference to the working session (that is what `SSL_get_session` does). Users who want to retain a copy of the session should always use `wolfSSL_get1_session` as that will guarantee either an incremented refCount of a duplicated object.

Reply:
This will break existing compatibility. Did you consider making the WOLFSSL_SESSION in WOLFSSL a pointer and allocating from heap? It would not increase memory use, since the WOLFSSL object is already allocated from the heap.

Reply:
>This will break existing compatibility.

While true, the assumption is that a pointer directly to the cache entry is just as unsafe. The client user has no way of knowing if the session didn't change before setting it in a new SSL object. We could make `wolfSSL_get_session` and `wolfSSL_get1_session` do the same thing but that will cause leaks since `wolfSSL_get_session` is meant to return a reference to the SSL object's working session.

>Did you consider making the WOLFSSL_SESSION in WOLFSSL a pointer and allocating from heap?

This is an interesting idea. It should be fairly easy to do. The only issue would be that the heap is further fragmented. Do you think I should pursue this direction?

Reply:
Yes please. You'll have to track that the allocated WOLFSSL_SESSION was retrieved and assume to not free it on `wolfSSL_free`. I think the session_ref stuff already does something similar, which could all be cleaned up and combined.

Reply:
Do you mean that the session should point into the cache? I was thinking it would be allocated on wolfSSL_new and free'd on wolfSSL_free. A user could retain the object by either calling wolfSSL_get1_session or the up_ref API.

Reply:
No, allocated.

Reply:
Got it. I'll try to get it done tomorrow.